frontend/packages/knative-plugin/src/models.ts

```diff
@@ -60,7 +60,7 @@ export const RouteModel: K8sKind = {
   labelPlural: 'Routes',
   plural: 'routes',
   id: 'route',
-  abbr: 'R',
+  abbr: 'RT',
```

Comment:
Route Details Page header doesn't have the `RT`:

![Screen Shot 2019-12-04 at 5 26 46 PM](https://user-images.githubusercontent.com/8126518/70187066-5ccf3400-16bb-11ea-9361-bec926ee2cd0.png)

Reply:
Yeah, that uses models retrieved from the APIs which include the same version but that model uses 'R'. Since the header is generic I can't find a decent way to fix this without a changes to how we handle models in general. 

Do we know if there is there an open issue to fix the k8s-model code to handle duplicate entries?

Reply:
> Do we know if there is there an open issue to fix the k8s-model code to handle duplicate entries?

@jeff-phillips-18 What is the bug exactly? I'm not aware of a problem.

Reply:
@spadgett it was an issue we encountered a while back (https://github.com/openshift/console/pull/2917#issuecomment-542732765), I logged https://jira.coreos.com/browse/ODC-2125 to track the problem. It since has grown beyond the badges having issues. `kind` lookup is problematic (ie [inject.js#L26-L28](https://github.com/openshift/console/blob/master/frontend/public/components/utils/inject.js#L26-L28)), since it doesn't validate different api groups. So the [model lookup](https://github.com/openshift/console/blob/master/frontend/public/module/k8s/k8s-models.ts#L29) has issues traversing the `kind`s.

I tried making a change to allow for CRDs to get special treatment, but it caused issues with lookup for the Networking Service and Route models. This PR has similar issues when it tries to pull the model from `allModels()` for Knative Routes or Services.

We'll likely have to do something about this in 4.4 if we want to cleanup the issues around these two types of models.

cc @christianvogt

Reply:
I'm still unclear on the problem. Can you point me to the code where this occurs? If you include the API group, it should just work.

We should really be using `connectToModels`, not `allModels`

Reply:
Ah, we probably should only fallback to `kind` if `!model.crd`

https://github.com/openshift/console/blob/master/frontend/public/module/k8s/k8s-models.ts#L30

Reply:
Yeah, I suspect it breaks here since `hasModel` will return true when we fallback to `kind`:

https://github.com/openshift/console/blob/master/frontend/public/module/k8s/k8s-models.ts#L34

Reply:
In a nutshell, I think the problem is [here](https://github.com/openshift/console/blob/master/frontend/public/module/k8s/k8s-models.ts#L29).

Some debugging...
```
k8sModels = k8sModels.withMutations((map) => {
  const pluginModels = _.flatMap(registry.getModelDefinitions().map((md) => md.properties.models));
  map.merge(modelsToMap(pluginModels.filter((model) => {
    const yesHasModel = hasModel(model);
    console.debug('Adding Model', referenceForModel(model), model, yesHasModel ? 'Already There' : 'Not there, adding');
    return !yesHasModel;
  })));
});
```
Output:
```
Adding Model (serving.knative.dev~v1beta1~Route)? Already there, ignoring
Adding Model (serving.knative.dev~v1beta1~Service)? Already there, ignoring
```

Reply:
Right, the `hasModel` check if wrong, causing that. It should be

```diff
diff --git a/frontend/public/module/k8s/k8s-models.ts b/frontend/public/module/k8s/k8s-models.ts
index 5a25caf57..23a28cbe4 100644
--- a/frontend/public/module/k8s/k8s-models.ts
+++ b/frontend/public/module/k8s/k8s-models.ts
@@ -27,7 +27,7 @@ export const modelsToMap = (models: K8sKind[]): ImmutableMap<K8sResourceKindRefe
 let k8sModels = modelsToMap(_.values(staticModels));

 const hasModel = (model: K8sKind) =>
-  k8sModels.has(referenceForModel(model)) || k8sModels.has(model.kind);
+  k8sModels.has(referenceForModel(model)) || (model.crd && k8sModels.has(model.kind));
```

Reply:
I'm not sure why we need the `hasModel` check that at all, though. I guess we're worried about a plugin overwriting a model from the core?

Reply:
Yeah, I don't think we can do [this](https://github.com/openshift/console/pull/3581#discussion_r354417629)

![image](https://user-images.githubusercontent.com/8126518/70254889-d1ee4800-1753-11ea-9916-dfcf32f9028d.png)

The Knative Service kicks in as the last one registered as the `kind: Service`

Reply:
It's a lookup problem I think more than a storage problem.

Reply:
yeah, i think we can fix that as well

Reply:
If we remove `kindObj` usage, absolutely. It's looking up the old way.

Reply:
This should address the problem with the services list. I'm still testing, though.

https://github.com/openshift/console/pull/3690/files#diff-5193194ffa917b7036f98babfc19dfa9R30